lib/voice/VoiceConnection.js

```diff
@@ -585,34 +639,61 @@ class VoiceConnection extends EventEmitter {
-                    /**
-                     * Fired to warn of something weird but non-breaking happening
-                     * @event VoiceConnection#warn
-                     * @prop {String} message The warning message
-                     */
-                    this.emit("warn", "Failed to decrypt received packet");
-                    return;
-                }
+            const rtpVersion = msg[0] >> 6;
+            if(rtpVersion !== 2) { // Check if RTP version is 2
+                this.emit("warn", "Recieved packet with an invalid RTP version: " + rtpVersion);
+                return;
             }
+
+            // RFC3550 5.1: RTP Fixed Header Fields
             const hasExtension = !!(msg[0] & 0b10000);
+            const hasPadding = !!(msg[0] & 0b100000);
             const cc = msg[0] & 0b1111;
-            if(cc > 0) {
-                data = data.subarray(cc * 4);
+            const nonce = Buffer.alloc(this.mode === "aead_aes256_gcm_rtpsize" ? 12 : 24);
+            msg.copy(nonce, 0, msg.length - 4, msg.length);
+
+            // Header Size = Fixed Header Length + (CSRC Identifier Length * CSRC count) + Extension
+            let headerSize = 12 + (cc * 4);
+            if(hasExtension) {
+                headerSize += 4;
+            }
+
+            let data;
+            if(this.mode === "aead_aes256_gcm_rtpsize") {
+                const decipher = crypto.createDecipheriv("aes-256-gcm", this.secret, nonce);
+                decipher.setAAD(msg.subarray(0, headerSize));
+                decipher.setAuthTag(msg.subarray(msg.length - 20, msg.length - 4));
+                data = Buffer.concat([decipher.update(msg.subarray(headerSize, msg.length - 20)), decipher.final()]);
+            } else if(Sodium) {
+                data = Buffer.alloc(msg.length - (headerSize + 4) - Sodium.crypto_aead_xchacha20poly1305_ietf_ABYTES);
+                Sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(
+                    data,
+                    null,
+                    msg.subarray(headerSize, msg.length - 4),
+                    msg.subarray(0, headerSize),
+                    nonce,
+                    this.secret
+                );
             }
+
+            // RFC3550 5.1: Padding (may need testing)
+            if(hasPadding) {
+                const paddingAmount = data.subarray(0, data.length - 1);
+                if(paddingAmount < data.length) {
+                    data = data.subarray(0, data.length - paddingAmount);
+                }
+            }
+
             // Not a RFC5285 One Byte Header Extension (not negotiated)
-            if(hasExtension) { // RFC3550 5.3.1: RTP Header Extension
-                const l = data[2] << 8 | data[3];
-                data = data.subarray(4 + l * 4);
+            if(hasExtension) {
+                const extOffset = headerSize - 4;
+                if(msg[extOffset + 0] == 0xbe && msg[extOffset + 1] == 0xde) { // RFC3550 5.3.1: RTP Header Extension
```

Comment:
use strict equality (this isn't (and wasn't) enforced by linting yet)
```suggestion
                if(msg[extOffset + 0] === 0xbe && msg[extOffset + 1] === 0xde) { // RFC3550 5.3.1: RTP Header Extension
```